stock_picking_propagate_scheduled_date/models/stock_move.py

```diff
@@ -2,32 +2,65 @@
 # @author Vincent Van Rossem <vincent.vanrossem@camptocamp.com>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
+from datetime import timedelta
+
 from odoo import fields, models
 
 
 class StockMove(models.Model):
     _inherit = "stock.move"
 
-    def _propagate_date(self, new_date):
-        """Propagate the date of a move to its destination."""
-        already_propagated_ids = self.env.context.get(
-            "date_propagation_ids", set()
-        ) | set(self.ids)
-        self = self.with_context(date_propagation_ids=already_propagated_ids)
+    def write(self, vals):
+        previous_dates = {
+            move.id: move.date for move in self.with_context(prefetch_fields=False)
+        }
+        res = super().write(vals)
+        if "date" in vals and not self.env.context.get(
+            "_propagate_date_no_recursion", False
+        ):
+            self._propagate_date_no_recursion(previous_dates, vals.get("date"))
+        return res
+
+    def _get_moves_in_down_chain(self, chained_dest_moves=None):
+        """Get all the destination moves in the down chain of a move recursively."""
+        self.ensure_one()
+        if chained_dest_moves is None:
+            chained_dest_moves = set()
+
+        # If the dest move is already checked,
+        # return the chained_dest_moves set to avoid infinite recursion
+        if self.id in chained_dest_moves:
+            return chained_dest_moves
+
+        # Add the current move to the chained_dest_moves set
+        chained_dest_moves.add(self.id)
+
+        # Recursively find all dest moves in the down chain
+        for move_dest in self.move_dest_ids:
+            move_dest._get_moves_in_down_chain(chained_dest_moves)
+
+        return chained_dest_moves
```

Comment:
Could you try using a recursive SQL query here? 

Also the chained move which are Done or Canceled should not be updated

Reply:
And possibly remove the ensure_one() to be able to work on a full recordset and perform a single SQL query.

Reply:
```
query = """
WITH RECURSIVE cte AS
(
SELECT
    r.move_orig_id, r.move_orig_id as from_id, r.move_dest_id as dest_id
FROM
    stock_move_move_rel r
WHERE r.move_orig_id in %s

UNION ALL

SELECT
    cte.move_orig_id,
    r.move_orig_id as from_id,
    r.move_dest_id as dest_id
FROM
    stock_move_move_rel r, cte
WHERE
    r.move_orig_id = cte.dest_id
)

SELECT 
   move_orig_id, 
   array_agg(dest_id)
FROM
   cte
group by move_orig_id
;"""
self.env.cr.execute(query, (tuple(self.ids),)
chained = dict(self.env.cr.fetchall()  # key: start move id, values: list of chained moves
```